src/app/clusters/temperature-control-server/temperature-control-server.cpp

```diff
@@ -0,0 +1,188 @@
+/**
+ *
+ *    Copyright (c) 2023 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ *
+ */
+
+#include <app-common/zap-generated/attributes/Accessors.h>
+#include <app-common/zap-generated/cluster-objects.h>
+#include <app/InteractionModelEngine.h>
+#include <app/clusters/temperature-control-server/supported-temperature-levels-manager.h>
+#include <app/util/attribute-storage.h>
+#include <app/util/error-mapping.h>
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::app::Clusters;
+using namespace chip::app::Clusters::TemperatureControl;
+using namespace chip::app::Clusters::TemperatureControl::Attributes;
+using namespace chip::DeviceLayer;
+using chip::Protocols::InteractionModel::Status;
+
+namespace {
+
+class TemperatureControlAttrAccess : public AttributeAccessInterface
+{
+public:
+    // Register for the TemperatureControl cluster on all endpoints.
+    TemperatureControlAttrAccess() : AttributeAccessInterface(Optional<EndpointId>::Missing(), TemperatureControl::Id) {}
+
+    CHIP_ERROR Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder) override;
+};
+
+} // namespace
+TemperatureControlAttrAccess gAttrAccess;
+
+CHIP_ERROR TemperatureControlAttrAccess::Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
+{
+    VerifyOrDie(aPath.mClusterId == TemperatureControl::Id);
+
+    TemperatureControl::SupportedTemperatureLevelsManager::Iterator iterator(aPath.mEndpointId);
```

Comment:
I'm not quite sure how this is supposed to work.  Who implements the `Next` for this iterator?  Where is the Label for the struct stored?

One option is to define a new struct type that inherits from `Structs::TemperatureLevelStruct::Type` and has storage for the label, and then have a way to iterate into _those_ which copies the label and points the span to the right place.

Another option is to define that the iterator itself can store state, but then we need to allow the iterator class to be provided by the actual manager implementation, right?

Reply:
Updated interface